File: day03/python/part1.py
#!/usr/bin/env python3

import helper


def get_column(i: int, lines: list[str]) -> list[str]:
    return [line[i] for line in lines]


def main():
    # lines = helper.read_lines("example.txt")
    lines = helper.read_lines("input.txt")
    # for line in lines:
        # print(line)
    no_of_columns = len(lines[0])
    # print(no_of_columns)
    result = []
    for i in range(no_of_columns):
        column = get_column(i, lines)
        zeros = column.count('0')
        ones = column.count('1')
        result.append('0' if zeros > ones else '1')
        # print("".join(column))
    #
    gamma_str = "".join(result)
    epsilon_str = gamma_str.replace('0', 'x').replace('1', '0').replace('x', '1')
    # print(gamma_str)
    # print(epsilon_str)
    gamma = int(gamma_str, 2)
    epsilon = int(epsilon_str, 2)

    print(gamma * epsilon)

##############################################################################

if __name__ == "__main__":
    main()

File: day03/julia/part1.jl
#!/usr/bin/env julia

using FromFile

@from "Helper.jl" import Helper

function get_column(i::Int, lines::Vector{String})::Vector{Char}
    return [line[i] for line in lines]
end

function main()
    # fname = "example.txt"
    fname = "input.txt"

    lines = Helper.read_lines(fname)
    no_of_columns = length(lines[1])

    result::Vector{Char} = []
    for i in 1:no_of_columns
        column = get_column(i, lines)
        zeros = count(c -> c == '0', column)
        ones = length(column) - zeros
        append!(result, zeros > ones ? '0' : '1')
    end
    gamma_str = join(result)
    epsilon_str = begin
        tmp = replace(gamma_str, '0' => 'x')
        tmp = replace(tmp, '1' => '0')
        replace(tmp, 'x' => '1')
    end
    gamma = parse(Int, gamma_str, base=2)
    epsilon = parse(Int, epsilon_str, base=2)
    println(gamma)
    println(epsilon)
    println("---")
    println(gamma * epsilon)
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
